src/inet/UDPEndPoint.cpp

```diff
@@ -543,6 +543,30 @@ INET_ERROR UDPEndPoint::SendMsg(const IPPacketInfo * pktInfo, System::PacketBuff
 
 #if CHIP_SYSTEM_CONFIG_USE_LWIP
 
+    System::PacketBufferHandle msgCopy;
+    pbuf * pmsg = NULL;
+
+    if (!msg.HasSoleOwnership())
+    {
+        // when retaining a buffer, the caller expects the msg to be unmodified.
+        // LwIP stack will normally prepend the packet headers as the packet traverses
+        // the UDP/IP/netif layers, which normally modifies the packet. We need to clone
+        // msg into a fresh object in this case, and queues that for transmission, leaving
+        // the original msg available after return.
+        msgCopy = msg.CloneData();
```

Comment:
It does not look like we need msgCopy.

How about:

```cpp
if (!msg.HasSoleOwnership()) {
   // comments...
   msg = msg.CloneData();
}
```

you will not need any other changes.
Also consider doing this for all platforms (not only LWIP) for consistency sake. It has a cost but it seems minimal.

Reply:
It take me some time to understand the ref count with this change.

1. With msgCopy, the ref count is easy to calculate since we didn't touch the original logic.
Suppose we pass a packet which has been retained, msg.ref ==2. and msgCopy.ref =1 
When the function is returned, the ref count of the original packet become 1 since msg is destructed and the ref count of the copied packet become 0 and the copied packet is freed.

2. Without msgCopy, the ref count is calculated as:
Suppose we pass a packet which has been retained, msg.ref ==2. 
After msg = msg.CloneData();
The ref count of the original packet become 1 since msg gave up the ownership and msg take the ownership of copied packet with ref count = 1. 
When the function is returned, the ref count of the original packet is still 1 and the ref count of the copied packet become 0  since msg is destructed. The copied packet is freed.

Am I right?

Reply:
I believe this is correct. @kpschoedel 

Generally I am assuming that packetbufferhandles have correct/safe reference counting and I am using them as regular values without trying to manually match reference counts. Their functionality should be 'has a reference count while object is alive, drops reference count when object is destructed'

Reply:
Yep, that's how handles work.

The one thing that might be a footgun in that sense is if you do `handle = std::move(handle)`.  That will lead to a dropped ref and a null pointer.  But `handle = handle.CloneData()` should works just fine.

Reply:
Checking for `handle = std::move(handle)` would cost around 100 bytes in the current tree. I think that's acceptable.

Reply:
Yes, only "msg = msg.CloneData()" could do the trick, "msg = std::move(msg.CloneData())" does not.

Reply:
`msg = std::move(msg.CloneData())` should be fine as well; it will end up behaving just like `msg = msg.CloneData()` because in both cases the RHS is an rvalue ref.

To be clear, I am not saying we need to check for `handle = std::move(handle)`.  I'm just saying that this is the one "may act like a reference, not a value" edge case.  In practice, no one is doing that, though, and it would be pretty obvious if someone did.